ts/boot/migrateToPagination.ts

```diff
@@ -0,0 +1,66 @@
+import * as Either from "fp-ts/lib/Either";
+
+import {
+  messagesStatusSelector,
+  MessageStatus
+} from "../store/reducers/entities/messages/messagesStatus";
+import { Store } from "../store/actions/types";
+import { removeMessages } from "../store/actions/messages";
+
+type Failure = {
+  error: unknown;
+  messageId: string;
+};
+
+type MigrationResult = {
+  failed: Array<Failure>;
+  succeeded: Array<string>;
+};
+
+export default async function init(
+  store: Store,
+  upsert: (id: string, messageStatus: MessageStatus) => Promise<void>
+): Promise<MigrationResult> {
+  const data = messagesStatusSelector(store.getState());
+  const allIds = Object.keys(data);
+  if (allIds.length < 1) {
+    return { failed: [], succeeded: [] };
+  }
+  // TODO: dispatch to show the loader
+
+  const requests: Array<Promise<Either.Either<Failure, string>>> = allIds.map(
+    async id => {
+      const messageStatus = data[id];
+      if (messageStatus) {
+        try {
+          await upsert(id, messageStatus);
+          return Either.right<Failure, string>(id);
+        } catch (error) {
+          return Either.left<Failure, string>({ error, messageId: id });
+        }
+      } else {
+        return Either.right<Failure, string>(id);
+      }
+    }
+  );
+
+  return Promise.all(requests).then(results => {
```

Comment:
How does `Promise.all` handle concurrency? I mean, does it start all the network requests at the same time?

Reply:
Promises are run in parallel in no specific order (although the order of the fulfilled results is guaranteed).
The gotcha here is that `Promise.all` fails immediately if any of the Promises throws.
What we want here instead is [Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) behavior which is unfortunately not supported at the moment.

That's why I [make sure to catch the errors](https://github.com/pagopa/io-app/pull/3841/files#diff-bcbf6afc3d95f5b125321b43a35962fc8b5fa5186b29df65796d739bd9113817R38) and I put [a specific test](https://github.com/pagopa/io-app/pull/3841/files#diff-67144522210873f6e25ec306e8dc26f17aa2b6626dd44f2c5a5b931db6ea53faR65) to ensure it works as expected.

Reply:
My fear is that if you have "a lot" of messages you're going to kill the network, having a high failure rate (because some request is inevitably going to timeout). It would be better to send them serially or (at least) in batches.

Reply:
Good point. Due to the sensitivity of the matter I won't share metrics here, but I encourage you to check the [migration strategy](https://pagopa.atlassian.net/wiki/spaces/IOAPP/pages/474349926/Rilascio+in+produzione) where the issue is being discussed. Your opinion there is welcome and valued!

Reply:
@pietro909 @emiliopavia If it's fine for you too, I think this can be merged and the batching behaviour handled in another ticket since it's not gonna be enabled right now 🙂. Let me know your thoughts!

Reply:
@dgopsq sure! Let's move things forward.